text/3437-implementable-trait-alias.md

```diff
@@ -0,0 +1,959 @@
+let signed_iter = [1_i32].into_iter();
+//IntIter::next(&mut signed_iter); // ERROR: Expected `<Self as Iterator>::Item` to be `u32`, it is `i32`
+```
+
+Implementable trait aliases can also be used with associated type bounds.
+
+```rust
+trait IteratorAlias = Iterator;
+let _: IteratorAlias<Item = u32> = [1_u32].into_iter();
+
+trait IntIter = Iterator<Item = u32> where Self: Clone;
+let _: IntIter<Item = u32> = [1_u32].into_iter(); // `Item = u32` is redundant, but allowed
+//let _: IntIter<Item = f64> = [1.0_f64].into_iter(); // ERROR: `Item = f64` conflicts with `Item = u32`
+```
+
+Items from traits in `where` clauses of the alias are accessible, unless
+shadowed by items in the primary trait(s):
+
+```rust
+trait Foo {
+    fn frob();
+    fn frit();
+}
+
+trait Bar {
+    fn frit();
+    fn bork();
+}
+
+trait FooBar = Bar where Self: Foo;
+
+fn example<T: FooBar>() {
+    T::frob(); // resolves to `<T as Foo>::frob`
+    T::frit(); // resolves to `<T as Bar>::frit`
+    T::bork(); // resolves to `<T as Bar>::bork`
+}
+```
+
+## Aliases with multiple primary traits
+
+A trait alias with multiple primary traits can be implemented, unless one of the
+primary traits requires specifying an item that conflicts with an item of the
+same name in a different primary trait.
+
+```rust
+trait Foo {
+    fn frob();
+}
+
+trait Bar {
+    fn frob() {}
+}
+
+// This isn't implementable, due to conflict between `Foo::frob` and `Bar::frob`
+trait FooBar = Foo + Bar;
+```
+
+If the confliting items all have defaults, the alias will be implementable, but
+overriding the defaults will not be possible.
+
+```rust
+trait Foo {
+    fn frob() {}
+}
+
+trait Bar {
+    fn frob() {}
+}
+
+// This is implementable, but the `impl` block won't be able
+// to override the default bodies of the `frob()` functions.
+trait FooBar = Foo + Bar;
+```
+
+Name conflicts of this sort also cause ambiguity when using the alias:
+
+```rust
+fn example<T: FooBar>() {
+    T::frob(); // ERROR: ambiguous
+}
+```
+
+To resolve these conflicts, you can use trait alias bodies, as described below.
+
+## Bodies for trait aliases
+
+Trait aliases can now optionally contain a body, specifying aliases for various
+items. These can be types, constants, or functions.
+
+### `type`s and `const` items in trait alias bodies
+
+```rust
+trait Foo {
+    type Assoc;
+    const ASSOC: i32;
+}
+
+trait Alias = Foo {
+    type AssocVec = Vec<Self::Assoc>;
+    const ASSOC_PLUS_1: i32 = Self::ASSOC + 1;
+}
+```
+
+`<T as Alias>::AssocVec` means the same thing as `Vec<<T as Foo>::Assoc>`, and
+`<T as Alias>::ASSOC_PLUS_1` is equivalent to `const { <T as Foo>::ASSOC + 1 }`.
+
+To be implementable, a `type` or `const` alias item must obey certain
+restrictions. It must either be set equal to an item of a primary trait of the
+alias:
+
+```rust
+trait Foo {
+    type Assoc;
+    const ASSOC: i32;
+}
+
+trait Alias = Foo {
+    type Associated = Self::Assoc;
```

Comment:
are you required to write `type Associated: ?Sized = Self::Assoc` if you want to avoid a default `Sized` bound when defining a trait alias?

Reply:
No, there is not, just as there is no such default `Sized` bound for type aliases in other contexts.

Reply:
there *are* default bounds for associated types, which can be written with basically the same syntax, so I think this definitely needs to be specified.
```rust
trait Tr<T: Deref> {
    type Ty = <T as Deref>::Target; // has a Sized bound
}
trait Tr2<T: Deref> = Send {
    type Ty = <T as Deref>::Target; // has no Sized bound
}
```

Reply:
I’ve added a note in the RFC.